### ISSUE TITLE
Inconsistent Behavior with `readonly` in Template Literal Indexed Access Types

### ISSUE THREAD

**Issue by 13OnTheCode** [microsoft/TypeScript#58035]
### 🔎 Search Terms

"readonly" "template literal" "index signatures"

### 🕗 Version & Regression Information

v5.4.3

### ⏯ Playground Link

https://www.typescriptlang.org/play?#code/C4TwDgpgBAogHgQwLZgDbQLxQN4CgpQDacAXFAAYAk2AzsAE4CWAdgOYC+AtMAo6uQF0ydJm3xFSFABYQEAE07URLDoOEMV4+rLkB7ZqhASy5GfMW0NbLjz5qoysQW3z9h4xREXHqoQ6usuOy4uHIQAMaoCNpQ4fp0UMAQdGTwyGgQIUl0hADkXlIIhbkCUFi5AIwATADMuVnJwHlmCoWF3LyoJWVQlbX1uKCQUAAqFT0ACozhANYAPGko6AA0UDMQILoAZrCISxAAfLgA9McEUAB6APy4QA

### 💻 Code

```ts
type Example = {
  [x: `${string}-tail`]: string
  [x: `head-${string}`]: string
  readonly [x: `head-${string}-tail`]: string
  readonly [x: `str-${string}`]: string
}

declare const test: Example

test['str-haha'] = '123' // Expected: error , Actual: error

test['head-haha-tail'] = '123' // Expected: error , Actual: no error

type T1 = Pick<Example, keyof Example>
// {
//    [x: `${string}-tail`]: string;
//    [x: `head-${string}`]: string;
//    [x: `head-${string}-tail`]: string;
//    readonly [x: `str-${string}`]: string;
// }
```

### 🙁 Actual behavior

I have encountered an inconsistent behavior when using `readonly` with template literal indexed access types in TypeScript. Specifically, when a string template index signature extends another string template index signature, attempts to mark it as `readonly` seem to be ignored in the resulting type when using `Pick<Type, keyof Type>`.

### 🙂 Expected behavior

The keyword `readonly` should be retained, or an error prompt should be given when adding it at the beginning.

**Comment by ahejlsberg** [comment]
This is working as intended. We consider an element access to be read-only when _all_ applicable index signatures have a `readonly` modifier. In the example, the string `"head-haha-tail"` matches three index signatures, two of which are not read-only. Therefore, the assignment is permitted. A similar thing happens when resolving the properties of the `Pick<Example, keyof Example>` mapped type. For each generated index signature we query the _combined_ read-only-ness of applicable index signatures in the source type, and `` `head-${string}-tail` `` matches three, two of which aren't read-only.

**Comment by 13OnTheCode** [comment]
@ahejlsberg I appreciate your response.

Sorry I didn't make it clear before, I understand that the behavior is working as expected, but my confusion is:

The string index `head-${string}-tail` matches three index signatures, two of which are not read-only. So, when we try to add `readonly` to ``head-${string}-tail`` only, there should be an error prompt, but the actual situation seems to be correct.

![微信截图_20240414192355](https://github.com/microsoft/TypeScript/assets/137921275/1a83a73d-f560-415f-be63-324396b9977b)
